middleman_protocol/middleman_protocol/message.py

```diff
@@ -0,0 +1,196 @@
+from hashlib import sha256
+
+from construct import Byte
+from construct import Bytes
+from construct import Const
+from construct import Container
+from construct import Enum
+from construct import GreedyBytes
+from construct import Int16ub
+from construct import Int32ub
+from construct import Prefixed
+from construct import Struct
+from construct import VarInt
+
+from golem_messages.message.base import Message as BaseGolemMessage
+
+from .constants import FRAME_SEPARATOR
+from .constants import PayloadType
+from .constants import PAYLOAD_TYPE_TO_MIDDLEMAN_MESSAGE_CLASS
+from .exceptions import MiddlemanProtocolError
+
+
+class AbstractMiddlemanMessage:
+
+    __slots__ = [
+        'payload',
+        'payload_type',
+        'request_id',
+    ]
+
+    def __init__(self, payload, request_id):
+        if type(self) == AbstractMiddlemanMessage:
+            raise NotImplementedError(
+                'AbstractMiddlemanMessage cannot be instantiated directly.'
+            )
+
+        self._validate_request_id(request_id)
+        self.request_id = request_id
+
+        self._validate_payload(payload)
+        self.payload = self._serialize_payload(payload)
+
+    @classmethod
+    def factory(cls, payload_type, payload, request_id):
+        assert payload_type in PayloadType
+        assert hasattr(sys.modules[__name__], PAYLOAD_TYPE_TO_MIDDLEMAN_MESSAGE_CLASS[payload_type])
+
+        return getattr(
+            sys.modules[__name__],
+            PAYLOAD_TYPE_TO_MIDDLEMAN_MESSAGE_CLASS[payload_type]
+        )(payload, request_id)
+
+    @classmethod
+    def get_message_format(cls) -> Struct:
+        message_format = Struct(
+            frame_signature=Bytes(64),  # TODO: How we gonna serialize
+            signed_part_of_the_frame=Struct(
+                request_id=Int32ub,
+                payload_length=Int16ub,
+                payload_type=Enum(Byte, PayloadType),
+                payload=Prefixed(VarInt, GreedyBytes),
+            ),
+            frame_separator=Const(FRAME_SEPARATOR),
+        )
+        return message_format
+
+    @classmethod
+    def deserialize(cls, raw_message):
+        message_format = cls.get_message_format()
+
+        deserialized_message = message_format.parse(raw_message)
+
+        cls._validate_signature(deserialized_message)
+        cls._validate_length(deserialized_message)
+
+        deserialized_payload = getattr(
+            sys.modules[__name__],
+            PAYLOAD_TYPE_TO_MIDDLEMAN_MESSAGE_CLASS[
+                PayloadType[str(deserialized_message.signed_part_of_the_frame.payload_type)]
+            ]
+        )._deserialize_payload(deserialized_message.signed_part_of_the_frame.payload)
+
+        return deserialized_payload
+
+    @classmethod
+    def _deserialize_payload(cls, payload):
+        return payload
+
+    @classmethod
+    def _validate_length(cls, message):
+        if not len(message.signed_part_of_the_frame.payload) == message.signed_part_of_the_frame.payload_length:
```

Comment:
Payload is allowed to be longer than the length stated in the header. Extra bytes should be ignored.

That's because you would have to see at least `len(separator)` bytes after the message to make sure there's nothing extra and you might not receive them until the sender starts pushing another message. So to avoid either waiting or risking that the message will be retroactively declared invalid, we allow the extra bytes.

But if your assumption is that the one creating the message is responsible for discarding the extra bytes then this should be an `assert` not an `if`.

Reply:
I don't understand from where the extra bytes can come from. If it is a stream, and one party will put a whole message into the stream (with the assumption that message is correct, which means it was created using this class), where can extra bytes come from ? 

Also, I think it should stay as `if`, because it then validates that the message was not malformed. And if it was malformed, it is not a problem with the code that sends it (tests cover that), but with client (it is a kind of attack). And then custom exception should be raised.

Reply:
> I don't understand from where the extra bytes can come from. 

For the purposes of this discussion I'll use this format when talking about the raw, binary stream: `<frame sig>|<request ID>|<payload length>|<payload type>|<payload>`. And `###` as separator.

Let's say you're listening for messages and you receive this:
```
###3ef2bc|1|30|GOLEM_MESSAGE|<45 bytes>###
```
The header says that the payload should have 30 bytes but there's 45 before the separator. According to the validation above this message is invalid. There's not problem with that if you receive the next separator immediately. But what it does not come immediately? E.g. the sender always includes the separator at the beginning of a message? You can't process the message. You have to sit an wait for the beginning of the next message.

Let's say you receive this:
```
###3ef2bc|1|30|GOLEM_MESSAGE|<30 bytes>
```
and the stream stops. You don't know when more will come and if it will come at all. You have to stop and wait for the separator because without it you cannot decide whether the message is valid or not. If the next thing you receive is the separator - it's valid. If it's anything else - the payload is longer than 30 bytes and it's invalid.

To prevent this the protocol is designed in such a way that as soon as you receive the number of bytes stated as the payload length, you can stop receiving, process the message and then if you receive any extra bytes you can safely discard them and wait for the separator.

There should not be any extra bytes in normal circumstances but if we could assume that the client always behaves correctly we would not need validations in the first place.

> Also, I think it should stay as if, because it then validates that the message was not malformed. And if it was malformed, it is not a problem with the code that sends it (tests cover that), but with client (it is a kind of attack). And then custom exception should be raised.

Yes, but I said that it should be an assertion **if your assumption is that the one creating the message is responsible for discarding the extra bytes**. `assert` would simply document that assumption. I did not know whether that was actually your assumption or whether you just did not implement this part of the protocol.

I see now that it's the latter so an `if` is the right way to do it.

Reply:
Are we running production code with optimization (`-O` option) ? Because if yes, the `assert` will be omitted there.

Reply:
That's the whole point. Having an `assert` here is the same as having no validation. If you think this can fail in production due to anything else than a bug, you do need a validation - don't use `assert`.

Reply:
:+1: done